Adjustment of Robert's PR

ThatOneGoat opened the pull request:
Slightly edited version of Robert's PR [here](https://github.com/ProjectSidewalk/sidewalk-panorama-tools/pull/12) to use the new Project Sidewalk adminapi/panos endpoint in the Download Runner.

Also adds in the scripts used to regather missing pano metadata.

misaugstad commented:
Can you write out at least a short set of testing instructions to streamline the testing process?

ThatOneGoat commented:
> Can you write out at least a short set of testing instructions to streamline the testing process?

Yup, since I mostly modified the download runner, here are some instructions to test that out:

1. The overall download process is to first gather pano info for a certain city via the `adminapi/panos` endpoint. This is done in the `fetch_pano_ids_from_webserver()` function, which ends up returning a map of `pano_ids` to `(pano_width, pano_height)` tuples indicating the size of the gsv panorama for that pano_id. This will do it for all non-tutorial panoramas as well as panoramas that have image metadata (we will assume that if the image has None for the `pano_width, pano_height`, it's because the pano is no longer retrievable from GSV and we can't redownload said panorama. To test this part is working successfully, we should log the length of the outputted map and make sure it is equal to the number of rows outputted via SQL from the prod database when using the same filter conditions (`pano_id` does exist, `pano_id != 'tutorial'`, and `pano_width, pano_height` are defined for said `pano_id`). 
2. The next part is the scraping of the xml metadata for the panos, the pano tiles (and formation into the final combined panorama), and the formation of the pano depth data from the XML info. For testing purposes, I added code (commented out) that lets you take a random subset of the acquired pano metadata and scrape just that subset. 
3. Firstly, we should confirm that the number of total attempted pano downloads matches the subset size. This can be easily confirmed by looking at the outputted `gsv_panorama_id_log.csv` file. The number of entries should be equivalent to the subset size. Likewise, make sure the counts of `success, failed, skipped` all add to the size of the subset.
4. Make sure that for every success, the subfolder in the output folder for that pano has the entire pano jpg (confirm for at least a couple that the pano does look complete to ensure that our attempts at getting accurate sizes was not in vain), associated xml metadata, and depth data. Also confirm that the download time for a single entire panorama is around 6 seconds (what we've deemed as reasonable download time).

If all this goes well, we should be cooking. I think it'd be a good idea to use this test process on a couple cities, like Seattle, SPGG, CDMX, Columbia, etc. (we could just do all the cities if the subset size isn't too large).

There also is another "big" script, `CropRunner.py`. However, I decided not to touch Robert's change since, in general, I wasn't sure if it was well-designed given that it requires pano info from a CSV which might be better acquired from an endpoint. That being said, it's not really that important for the CV project (and we need to update the cropping logic with the one in the CV project later anyways), so I decided it was non-blocking and moved forward with the PR (I'll circle back to this shortly once we get the experiments on GCP running since then we'll have a lot of downtime in waiting for experiments to finish).

ThatOneGoat commented:
Also, there is a function called `fetch_pano_ids_csv()` in the `DownloadRunner.py` that can extract `pano_ids` for downloading from a CSV file, but that needs to be updated to also grab the `image_width, image_height` from the CSV as well. Since this is what we'll probably use in place of the `fetch_pano_ids_from_webserver()` function when we go through to redownload panos that can be redownloaded.

ThatOneGoat commented:
Here is the new usage for the `DownloadRunner.py` script:
```
usage: DownloadRunner.py [-h] [-c [C]] d s

positional arguments:
  d           sidewalk_server_domain - FDQN of SidewalkWebpage server to fetch pano list from, i.e. sidewalk-sea.cs.washington.edu
  s           storage_path - location to store scraped panos

optional arguments:
  -h, --help  show this help message and exit
  -c [C]      csv_path - location of csv from which to read pano metadata
```
Essentially, I just added an optional csv path parameter in order to support reading pano metadata from a CSV rather than pulling from the webserver

misaugstad commented:
Okay I found a couple small issues. The first is that if an image has already been downloaded, but is being skipped, it's being marked as a 'success' in the logs instead of a 'skip'.

The second is that it looks like we are only trying to download panos if `image_width`/`image_height` are non-null from the `/adminapi/panos` endpoint. It isn't unreasonable to assume that that value will be filled in going forward, but the fact of the matter is that we do have data right now where that isn't filled in, and I'd like for us to be trying to download it, and marking that it's failing to download. So we should be trying anyway.

I realized that this was an issue because we somehow forgot to re-download the data for Chicago :upside_down_face:. We also did not update the metadata in Chicago, so the scraper was ignoring all of the data in Chicago. @ThatOneGoat can you send me those CSVs so that I can re-download Chicago data and update the metadata in the database. Then also make it so that we are trying to download regardless of whether or not image width/height are null. Here is the offending code:
```
    for value in jsondata:
        pano_id = value["gsv_panorama_id"]
        if pano_id not in unique_ids:
            # Check if the pano_id is an empty string
            if pano_id and pano_id != 'tutorial':
                # Check if image_width and image_height are non-None
                if value["image_width"] and value["image_height"]:
                    unique_ids.add(pano_id)
                    pano_info.append(value)
                else:
                    continue#print(f'Image width and height unknown for {pano_id}')
            else:
                print("Pano ID is an empty string or is for tutorial")
        else:
            print("Duplicate pano ID")
    assert len(unique_ids) == len(pano_info)
    return pano_info
```

misaugstad commented:
Oh I also think that when we download some XML that just says that the pano doesn't exist, we are logging that as a success as well. I'd like to log that as a fail. I think it's okay that we are saving that XML that says that the pano doesn't exist and are therefore skipping it on future runs... But I don't think that we should log it as a success.

misaugstad commented:
>if an image has already been downloaded, but is being skipped, it's being marked as a 'success' in the logs instead of a 'skip'.

I think that this may only be happening if it's marked as a success in the `gsv_panorama_id_log.csv` file. I just deleted that file for the SPGG server, and it went from 33,028 successes and 0 fails or skips for each of the past 5 days, to 33,028 SKIPS and no successes or fails.

misaugstad commented:
>Oh I also think that when we download some XML that just says that the pano doesn't exist, we are logging that as a success as well. I'd like to log that as a fail. I think it's okay that we are saving that XML that says that the pano doesn't exist and are therefore skipping it on future runs... But I don't think that we should log it as a success

I did not provide examples last week :) This is happening with panos that are provided by people other than Google. So we are getting a webpage with a 404 error as a response (as opposed to just empty XML), and we are just saving that HTML into an XML file. One example pano is this one: `CAoSLEFGMVFpcFBxd2ZyZTRBMVdkd1lNYmNveUEwU0U1c3p6SFBHN05PY21EcHot`

misaugstad commented:
I have now resolves all the issues mentioned above, as well as a few others I found in the meantime. I also updated the README to focus on using Docker, since that's much easier and it's how we use it ourselves! I've tested locally and on our production servers, so this should be good to merge now!

As noted in the README, the cropper hasn't really been tested... But I know that we have been working on an improved cropper in another repository. This should eventually be brought into this repository, but that's for a future PR :)
